File: spectral/spectral_code_challenge.py
# unsure if this is correct

import numpy as np
import matplotlib.pyplot as plt

# Create signal
srate = 1000
time = np.arange(-3, 3.00001, 1/srate)
pnts = len(time)
freqmod = np.exp(-time**2)*10+10
freqmod = freqmod + np.linspace(0, 10, pnts)
signal = np.sin(2*np.pi * ((time + np.cumsum(freqmod))/srate))

# Plot the signal
fig, axs = plt.subplots(3, 1, figsize=(8, 6))

axs[0].plot(time, signal, linewidth=1)
axs[0].set_xlabel('Time (s)')
axs[0].set_title('Time-domain signal')

n = 500
hz = np.linspace(0, srate, n)
tf = np.zeros((int(pnts/n)-1, len(hz)))
tv = np.zeros(int(pnts/n)-1)

for i in range(int(pnts/n)-1):
    datasnip = signal[i*n:(i+1)*n]
    pw = np.abs(np.fft.fft(datasnip))**2 / n  # Normalization
    tf[i, :len(hz)] = pw[:len(hz)]
    tv[i] = np.mean(time[i*n:(i+1)*n])

axs[1].imshow(tf.T, aspect='auto', origin='lower', extent=[tv.min(), tv.max(), 0, 40], cmap='hot', vmin=0, vmax=np.max(tf)/5)
axs[1].set_xlabel('Time (s)')
axs[1].set_ylabel('Frequency (Hz)')
axs[1].set_title("Professor's Solution")

# My Solution
import scipy.io

data = scipy.io.loadmat('spectral\spectral_codeChallenge.mat')
signal = data['signal'].flatten()

n = len(signal)
winlength = int(srate / 2)
step = int(winlength / 2)  # 50% overlap
td_wlen_start = np.arange(0, n-winlength, step)  # Time start of each window

hzW = np.linspace(0, srate/2, int(winlength/2))
signalpowW = []

for wi in range(len(td_wlen_start)):
    datachunk = signal[td_wlen_start[wi]:td_wlen_start[wi]+winlength]
    tmppow = np.abs(np.fft.fft(datachunk)/winlength)**2
    signalpowW.append(tmppow[:int(winlength/2)])

signalpowW = np.array(signalpowW).T

# Create a time vector based on the window start times
time_vector = (td_wlen_start) / srate

axs[2].imshow(signalpowW, aspect='auto', origin='lower', extent=[time_vector.min(), time_vector.max(), 0, 40], cmap='viridis', vmin=0, vmax=np.max(signalpowW)/5)
axs[2].set_xlabel('Time (s)')
axs[2].set_ylabel('Frequency (Hz)')
axs[2].set_title('My Solution')
fig.colorbar(axs[2].images[0], ax=axs[2])

plt.tight_layout()
plt.show()

File: spectral/spectral_code_challenge.m
clc;

% create signal
srate = 1000;
time  = -3:1/srate:3;
pnts  = length(time);
freqmod = exp(-time.^2)*10+10;
freqmod = freqmod + linspace(0,10,pnts);
signal  = sin( 2*pi * ((time + cumsum(freqmod))/srate) );


% plot the signal
figure(1), clf
subplot(3,1,1)
plot(time,signal,'linew',1)
xlabel('Time (s)')
title('Time-domain signal')


n  = 500;
hz = linspace(0,srate,n+1);
tf = zeros(floor(pnts/n)-1,length(hz));
tv = zeros(floor(pnts/n)-1,1);

for i=1:floor(pnts/n)-1
    datasnip = signal(i*n:(i+1)*n);

    pw = abs(fft(datasnip)).^2;
    tf(i,1:length(hz)) = pw(1:length(hz));
    tv(i) = mean(time(i*n:(i+1)*n));
end

subplot(3,1,2)
imagesc(tv,hz,tf')
axis xy
set(gca,'ylim',[0 40])
colormap hot
xlabel('Time (s)'), ylabel('Frequency (Hz)')
title("Professor's Solution")

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%% MY SOLUTION %%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

load spectral_codeChallenge.mat;

n = length(signal);
winlength = srate / 2;
td_wlen_start = 1:winlength:n-winlength; % Time start of each window

hzW = linspace(0, srate/2, floor(winlength));
signalpowW = [];

% hannw = 0.5 - cos(2*pi*linspace(0,1,winlength))./2;
for wi = 1:length(td_wlen_start)
    datachunk = signal(td_wlen_start(wi):td_wlen_start(wi)+winlength-1);
    % datachunk = datachunk.*hannw;
    tmppow = abs(fft(datachunk)/winlength).^2;
    signalpowW = [signalpowW, tmppow'];
end

% Create a time vector based on the window start times
time_vector = (td_wlen_start - 1) / srate;

subplot(3,1,3);
imagesc(time_vector, hzW, signalpowW);
axis xy;
ylim([0 40]);
xlabel('Time (s)');
ylabel('Frequency (Hz)');
title('My Solution');
colorbar;
